Make "~/.ansible" path configurable

htol opened the pull request:
##### SUMMARY
I want to control ~/.ansible directory location but it's ridiculously complex. My main motivation is clean and controllable $HOME directory.

Currently I have to setup around 24 environment variables to make it possible. And even if achieve it, it is not guaranteed that behavior will not be changed in the future or new plugin will not require to fix the path.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
base.yml

Fixes ansible/ansible#75788

htol committed:
make '~/.ansible' configurable by ANSIBLE_HOME

htol committed:
replace hardcoded '~/.ansible' to C.ANSIBLE_HOME

htol committed:
replace hardcoded '~/.ansible' to C.ANSIBLE_HOME

sivel commented:
Duplicate of https://github.com/ansible/ansible/pull/72043

sivel commented:
> Do I have any better way to collaborate?

1. Try to interact on the existing PR.  
2. Seek out interaction in #ansible-devel on IRC with the author
3. Come to an ansible core IRC meeting

fey commented:
Hi.
What's left to do for pr to be accepted?

bcoca commented:
change log, tests and passing existing tests

oaken-source commented:
any way I can help here? I too would like this feature to make it into ansible. from what I can see from the failed tests, it seems to be mostly setup errors in the tests themselves, not any significant failures.

htol commented:
Actually I have a time for continue but don't really understand next step. Can I get a clarification or direction?

stefanwascoding commented:
/rebuild_failed

stefanwascoding committed:
replace use of `get_config_value()` with constants where necessary

https://github.com/ansible/ansible/pull/77322#discussion_r835286830

stefanwascoding committed:
rename previously existing env ANSIBLE_HOME in env-setup script

Commit 6219fc4932797f29d23d331bbdaa6b3eed24f230 added an ANSIBLE_HOME variable to the `hacking/env-setup` script. Renamed that (`export`-ed) variable to ANSIBLE_DEV_HOME (as suggested by Brian Coca in https://github.com/ansible/ansible/pull/77322#issuecomment-1074353078) so as not to potentially mess with the new code.

stefanwascoding committed:
modify cache dir monkeypatching in galaxy api unit tests

since we are now using a different way to access the config variable in galaxy/api.py, we also need to update the monkey patch the tests are using to override the value

stefanwascoding committed:
update "version_added" to 2.14 for ANSIBLE_HOME

stefanwascoding commented:
@htol I've continued working on your PR in #77322 - mainly fixing the failing existing tests.  I would be happy for you to merge those changes back into this PR, as it seems wasteful to have two PRs for the same feature.

As far as I understand, what is missing is:
- [ ] adding documentation
- [ ] add tests for the new functionality.
- [ ] updating the change log

If you could give me write access to your fork I'd be happy to also continue working on this (on your PR).
Or, let me know if you're okay with me taking the lead on this, and I'll add you as collaborator to my fork.

htol commented:
@stefanwascoding I've added you as a collaborator to repository. Let me know please if it not enough to make changes in PR. Thank you!

stefanwascoding commented:
@htol Works.  Thanks.

The currently failing tests (Incidental AWS Python *) are temporary CI issues.

bcoca commented on lib/ansible/plugins/loader.py:
```diff
@@ -1145,7 +1145,7 @@ def _configure_collection_loader():
         return
 
-    finder = _AnsibleCollectionFinder(C.config.get_config_value('COLLECTIONS_PATHS'), C.config.get_config_value('COLLECTIONS_SCAN_SYS_PATH'))
+    finder = _AnsibleCollectionFinder(C.COLLECTIONS_PATHS, C.COLLECTIONS_SCAN_SYS_PATH)
```
don't make this change, we eventually want to phase out the constants in favor of the function calls.

sivel replied:
afaict only the constants are currently templated, `get_config_value` doesn't get templated.

stefanwascoding replied:
@bcoca as @sivel said, the `get_config_value` is not handling templates, only the constants do. See the [screenshot](https://user-images.githubusercontent.com/2292035/160599914-887296a0-a3ba-49eb-81f2-b3b0f4246f2b.png) on https://github.com/ansible/ansible/pull/77322#issuecomment-1081743289.

Personally, I would prefer to keep this change as is, and then open a separate PR, once this one is accepted, to rewrite the access of the configuration settings by providing a method (possibly by extending `get_config_value`) that is both dynamic like `get_config_value`, but also supports templating like the constants. Separate PR, because to me the modification (and code-wide usage update) of the config access seems to be too large to role into the PR about getting ANSIBLE_HOME working.

bcoca replied:
true, had meant to fix that a  while ago #77483

stefanwascoding replied:
👍🏻  In that case, I'll wait for the PR to land in devel, rebase, and switch (back) to `get_config_value` in all places.

bcoca replied:
no need to block on that

stefanwascoding commented:
PR changeset currently says it [affects 2.14](https://github.com/ansible/ansible/pull/76114/files#diff-3f3fbfd37f8f95b3e5f26617feb866a00e1cddb44fad1522321927d42349b877R15), but there's an `affects_2.13` label on the PR -- might be a holdover from when the PR was created last October.

bcoca commented on lib/ansible/config/base.yml:
```diff
@@ -193,10 +205,10 @@ COLLECTIONS_PATHS:
     and you want to add ``my.collection`` to that directory, it must be saved as
-    ``~/.ansible/collections/ansible_collections/my/collection``.
-  default: ~/.ansible/collections:/usr/share/ansible/collections
+    ``${ANSIBLE_HOME}/collections/ansible_collections/my/collection``.
```
I would not document as an 'env var' as it might confuse the user on how this works

htol replied:
Is it a better variant?
```
``'{{ ANSIBLE_HOME ~ "/collections" }}'``
```

bcoca replied:
yep

htol replied:
fixed

bcoca reviewed (changes requested):
code lgtm, just minor doc issue

htol committed:
fix description of collections with proper use of ANSIBLE_HOME
